examples/step-45/step-45.cc

```diff
@@ -14,417 +14,728 @@
+                                               constraints);
+      VectorTools::interpolate_boundary_values (dof_handler,
+                                                0,
+                                                BoundaryValues<dim>(),
+                                                constraints,
+                                                fe.component_mask(velocities));
+      VectorTools::interpolate_boundary_values (dof_handler,
+                                                1,
+                                                BoundaryValues<dim>(),
+                                                constraints,
+
+                                                fe.component_mask(velocities));
+// At this point we need to think about how we want to prescribe
+// periodicity. A face $face_2$  on the left boundary should be matched to
+// a face $face_1$ on the lower boundary given by $face_2=R\cdot face_1+b$
+// where the rotation matrix $R$ and the offset $b$ are given by
+// @f{align*}
+// R=\begin{pmatrix}
+// 0&1\\-1&0
+// \end{pmatrix},
+// \quad
+// b=\begin{pmatrix}0&0\end{pmatrix}.
+// @f}
+// These two objects not only describe how faces should be matched but also
+// in which sense the solution should be transformed from $face_2$ to
+// $face_1$.
+      FullMatrix<double> rotation_matrix(dim);
+      rotation_matrix[0][1]=1.;
+      rotation_matrix[1][0]=-1.;
+
+      Tensor<1,dim> offset;
+
+// For setting up the constraints, we first store the periodicity
+// information in an auxiliary object of type
+// std::vector<GridTools::PeriodicFacePair<typename
+// DoFHandler<dim>::cell_iterator> >. The periodic boundaries have the
+// boundary indicators 2 (x=0) and 3 (y=0). All the other parameters we
+// have set up before. In this case the direction does not matter. Due to
+// $face_2=R\cdot face_1+b$ is exactly what we want.
+      std::vector<GridTools::PeriodicFacePair<typename DoFHandler<dim>::cell_iterator> >
+      periodicity_vector;
+
+      const unsigned int direction = 1;
+
+      GridTools::collect_periodic_faces(dof_handler, 2, 3, direction,
+                                        periodicity_vector, offset,
+                                        rotation_matrix);
+
+// Next we need to provide information on which vector valued components of
+// the solution should be rotated. Due to the fact that in this example
+// there is just the velocity and this starts at the first component of the
+// solution vector we simply insert a 0:
+      std::vector<unsigned int> first_vector_components;
+      first_vector_components.push_back(0);
+
+// After setting up all the information in periodicity_vector all we have
+// to do is to tell make_periodicity_constraints to create the desired
+// constraints.
+      DoFTools::make_periodicity_constraints<DoFHandler<dim> >
+      (periodicity_vector, constraints, fe.component_mask(velocities),
+       first_vector_components);
```

Comment:
The intention at this point is to enforce periodic boundary conditions for the velocity...

Further, this component mask is never ignored.

The point is simply that enforcing periodic boundary conditions for the pressure is not needed, but also doesn't harm[1] (the closure of C^\infty and C^\infty_\per with respect to the L2 norm yields the same space, L2).

[1] In case of enough regularity, the pressure obviously has to fulfill the periodicity condition...